src/zkevm_specs/evm/execution/calldataload.py

```diff
@@ -0,0 +1,56 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ..table import RW, CallContextFieldTag, TxContextFieldTag
+from ..util import BufferReaderGadget
+from ...util.param import N_BYTES_WORD
+
+
+def calldataload(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.CALLDATALOAD)
+
+    # offset is the 64-bit offset to start reading 32-bytes from start of calldata.
+    offset = instruction.rlc_to_fq_exact(instruction.stack_pop(), n_bytes=8)
+
+    tx_id = instruction.call_context_lookup(CallContextFieldTag.TxId, RW.Read)
+
+    if instruction.curr.is_root:
+        calldata_length = instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataLength)
+        calldata_offset = 0
+    else:
+        calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+        calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+
+    src_addr = offset + calldata_offset
+    src_addr_end = calldata_length + calldata_offset
+
+    buffer_reader = BufferReaderGadget(
+        instruction, N_BYTES_WORD, src_addr, src_addr_end, N_BYTES_WORD
+    )
```

Comment:
I'm not sure about this part TBH.

As per Geth impl: https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L283-L292

At no point we require to read from the context CALLDATALENGTH or CALLDATAOFFSET.

The paper and the impl just say try to get the 32 bytes starting at offset. If not enough, right pad for each one.

Reply:
Might this be a result of some copying from CALLDATACOPY?

Reply:
I agree with @CPerezz. No need to lookup call_context for CallDataLength and CallDataOffset

Reply:
After listening to @ed255 I agree with what he said. I missed the fact that:

>after a call, the call data is copied from the caller memory into a buffer at scope.Contract.Input, but we don't do that
>[author=ed255]

Reply:
>in geth, after a call, the call data is copied from the caller memory into a buffer at scope.Contract.Input,   

@ed255  appreciate if you can share the code snippet doing this  . I only found `Contract.Input` is assigned here https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L164

Reply:
> > in geth, after a call, the call data is copied from the caller memory into a buffer at scope.Contract.Input,
> 
> @ed255 appreciate if you can share the code snippet doing this . I only found `Contract.Input` is assigned here https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L164

Sure!  Here's an example for the `CALL`:
1. `args` has a `[]byte` slice with the call data here: https://github.com/ethereum/go-ethereum/blob/538a86838423add89a0734d5334599898be91e5a/core/vm/instructions.go#L677 (technically it's not a copy of the call data, as in go, a slice is like a view into a possibly bigger array). This is passed to `interpreter.evm.Call` a few lines below.
2. In `interpreter.evm.Call`, the `args` is received as an argument called `input` and is passed to a function `evm.interpreter.Run`: https://github.com/ethereum/go-ethereum/blob/538a86838423add89a0734d5334599898be91e5a/core/vm/evm.go#L228
3. In `evm.interpreter.Run` we reach the line you mentioned, where `input` is assigned to `contract.Input` which is the line you shared.

So that's how the call data is set in the value `contract.Input`.